File: Data Structures & Algorithms/three-integer-sum/submission-1.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        //{-4,-1,-1,0,1,2}

        vector<vector<int>>output;

        int right,left;


        for(int i = 0;i<nums.size();i++){
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            right = nums.size()-1;
            left = i+1;
            while(left<right){
                if(nums[right]+nums[left]+nums[i]==0){
                    output.push_back({nums[i],nums[left],nums[right]});
                    while(left<right&&nums[left]==nums[left+1])left++;
                    while(left<right&&nums[right]==nums[right-1])right--;
                    left++;
                    right--;
                }else if(nums[right]+nums[left]+nums[i]>0){
                    right--;
                }else{
                    left++;
                }
        }}
        return output;
    }
};

File: Data Structures & Algorithms/house-robber-ii/submission-1.java
class Solution {
    HashMap<Integer,Integer> map = new HashMap<>();
    public int rob(int[] nums) {
        if(nums.length==1){return nums[0];}

        int temp = nums[nums.length-1];
        nums[nums.length-1] = 0;
        int left = helper(nums,0);
        map.clear();

        nums[nums.length-1] = temp;
        int right = helper(nums,1);

        return Math.max(right,left);
        
    }
    public int helper(int[] nums,int i){
        if(map.containsKey(i)){
            return map.get(i);
        }

        if(i >= nums.length){return 0;}
        int max = Math.max(helper(nums,i+2)+nums[i],helper(nums,i+1));

        map.put(i,max);
        return max;
    }
}
